dags/als-kp-dataset-dag.py

```diff
@@ -219,49 +220,171 @@ def transform_data(data: Dict[str, Any], **context) -> List[Dict[str, Any]]:
-           - Adds the Dataset to the collection
-        3. Stores the collection in Synapse
+        1. Retrieves the current C-PATH datasets.
+        2. Identifies and flags new datasets that are duplicates based on their 'title'.
 
         Arguments:
-            transformed_items: List of transformed and validated items
+            transformed_items (List[Dict[str, Any]]):
+                A list of transformed and validated items from the previous task.
             **context: Airflow task context containing DAG parameters
 
+        Raises:
+            ValueError: If any 'sameAs' or 'title' values are missing or empty after transformation.
+
         Returns:
-            str: The Synapse ID of the created and stored dataset collection
+            List[Dict[str, Any]]: A list of duplicate dataset items if there are any.
         """
         syn_hook = SynapseHook(context["params"]["synapse_conn_id"])
         synapse_client = syn_hook.client
 
-        columns = [
-            Column(name="id", column_type=ColumnType.ENTITYID),
-            Column(name="title", column_type=ColumnType.STRING, maximum_size=200),
-            Column(name="creator", column_type=ColumnType.STRING, maximum_size=100),
-            Column(name="keywords", column_type=ColumnType.STRING, maximum_size=250),
-            Column(name="subject", column_type=ColumnType.STRING, maximum_size=100),
-            Column(name="collection", column_type=ColumnType.STRING, maximum_size=100),
-            Column(name="publisher", column_type=ColumnType.STRING, maximum_size=100),
-            Column(name="species", column_type=ColumnType.STRING, maximum_size=100),
-            Column(name="sameAs", column_type=ColumnType.STRING, maximum_size=100),
-            Column(name="source", column_type=ColumnType.STRING, maximum_size=100),
-            Column(name="url", column_type=ColumnType.STRING, maximum_size=100),
-        ]
+        # Get current datasets
+        collection_id = context["params"]["collection_id"]
+        query_str = (
+            f"SELECT * FROM {collection_id} WHERE publisher='Critical Path Institute'"
+        )
 
-        dataset_collection = DatasetCollection(
-            name=context["params"]["collection_name"],
-            description=context["params"]["collection_description"],
-            parent_id=context["params"]["project_id"],
-            include_default_columns=True,
-            columns=columns,
+        current_data = synapse_client.tableQuery(query_str).asDataFrame()
+
+        current_datasets = list(current_data["sameAs"])
+
+        # Flag duplicates from new data
+        # Datasets are considered as "duplicates" if they have the same titles
+        duplicates = []
+        seen = {}
+
+        for item in transformed_items:
+            cpath_key = item.get("sameAs")
+            title = item.get("title")
+
+            if not cpath_key:
+                raise ValueError(f"Missing or empty 'sameAs' in item: {item}")
```

Comment:
Does this stop the pipeline?

Reply:
Yeah it does stop the pipeline. However, the pipeline should likely be stopped if that key is missing. According to the mapping defined [here](https://github.com/amp-als/data-model/blob/main/mapping/cpath.jsonata), this key is expected to always be present because the C-Path ID should always be present. If a dataset doesn't have an ID, we should probably investigate what's going on.